scipy/stats/_axis_nan_policy.py

```diff
@@ -232,6 +294,9 @@ def axis_nan_policy_wrapper(*args, _no_deco=False, **kwds):
             nan_policy = kwds.pop('nan_policy', 'propagate')
             del args  # avoid the possibility of passing both `args` and `kwds`
 
+            # convert masked arrays to regular arrays with sentinel values
+            samples, sentinel = _masked_arrays_2_sentinel_arrays(samples)
+
             if axis is None:
                 samples = [sample.ravel() for sample in samples]
                 axis = 0
```

Comment:
In an immediate followup, I'll:

- Move `_broadcast_arrays` and related functions from `_hypotests.py` into this file, consolidating them with `_broadcast_shapes_remove_axis`, etc.
- broadcast the `samples` together after the call to `_masked_arrays_2_sentinel_arrays` above. This will avoid trouble below when `axis` is bigger than the number of dimensions of the smallest array. 
- validate `axis` against the number of dimensions of the broadcasted arrays (currently `axis` can be any integer for 1D arrays, whereas it really should raise an error unless it's 0 or -1).
- add support for axis tuples as follows
```python3
# move axes in `axis` to the end, then ravel them
samples = [np.movedim(sample, axis, range(-len(axis), 0) 
               for sample in samples]
samples = [sample.reshape(sample.shape[:-len(axis)] + (-1,))
               for sample in samples]
axis = -1  # work over the last axis
```